File: year2/semestrul2/MPP/Laboratoare/lab4recuperare/src/main/java/org/example/Repository/DBParticipantRepository.java
package org.example.Repository;

import org.example.Model.Participant;
import org.example.Repository.Interfaces.ParticipantRepository;
import org.example.Repository.Utils.DBUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class DBParticipantRepository implements ParticipantRepository {
    private DBUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public DBParticipantRepository(Properties properties) {
        dbUtils = new DBUtils(properties);
    }

    @Override
    public List<Participant> findAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Participant> participantList = new ArrayList<>();
        try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * from Participant")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Participant participant = getParticipantFromResultSet(resultSet);
                    participantList.add(participant);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error finding all elements DB" + e);
        }
        logger.traceExit(participantList);
        return participantList;
    }

    @Override
    public List<Participant> findByNameandSurname(String name, String surname) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Participant> participantList = new ArrayList<>();
        try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Participant WHERE nume = ? AND prenume = ?")) {
            preparedStatement.setString(1, name);
            preparedStatement.setString(2, surname);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Participant participant = getParticipantFromResultSet(resultSet);
                    participantList.add(participant);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error finding Participant by name and surname: " + e);
        }
        logger.traceExit(participantList);
        return participantList;
    }

    @Override
    public List<Participant> findByEmail(String email) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Participant> participantList = new ArrayList<>();
        try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Participant WHERE email = ?")) {
            preparedStatement.setString(1, email);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Participant participant = getParticipantFromResultSet(resultSet);
                    participantList.add(participant);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error finding Participant by email: " + e);
        }
        logger.traceExit(participantList);
        return participantList;
    }

    private Participant getParticipantFromResultSet(ResultSet resultSet) throws SQLException {
        String nume = resultSet.getString("nume");
        String prenume = resultSet.getString("prenume");
        String varsta = resultSet.getString("varsta");
        String email = resultSet.getString("email");
        return new Participant(nume, prenume, varsta, email);
    }

    public Participant saveParticipant(Participant participant) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        if (con == null) {
            logger.error("Database connection is null");
            return null;
        }
        try (PreparedStatement preparedStatement = con.prepareStatement("INSERT INTO Participant (nume, prenume, varsta, email) VALUES (?, ?, ?, ?)")) {
            preparedStatement.setString(1, participant.getNume());
            preparedStatement.setString(2, participant.getPrenume());
            preparedStatement.setString(3, participant.getVarsta());
            preparedStatement.setString(4, participant.getEmail());
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error saving Participant: " + e);
        }
        logger.traceExit(participant);
        return participant;
    }
}

File: year2/semestrul2/MPP/Laboratoare/AvaloniaApplication1/Repository/DB/DbParticipantRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using AvaloniaApplication1.Models;
using AvaloniaApplication1.Repositories.Interfaces;
using AvaloniaApplication1.Repositories.Utils;

namespace AvaloniaApplication1.Repositories.Db
{
    public class DbParticipantRepository : IParticipantRepository
    {
        private readonly DbUtils _dbUtils;

        public DbParticipantRepository(DbUtils dbUtils)
        {
            _dbUtils = dbUtils;
        }

        public List<Participant> FindAll()
        {
            var list = new List<Participant>();
            using var connection = _dbUtils.GetConnection();
            connection.Open();

            using var cmd = connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM Participant";

            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                list.Add(ReadParticipant(reader));
            }

            return list;
        }

        public List<Participant> FindByNameAndSurname(string name, string surname)
        {
            var list = new List<Participant>();
            using var connection = _dbUtils.GetConnection();
            connection.Open();

            using var cmd = connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM Participant WHERE nume = @name AND prenume = @surname";
            var p1 = cmd.CreateParameter(); p1.ParameterName = "@name"; p1.Value = name; cmd.Parameters.Add(p1);
            var p2 = cmd.CreateParameter(); p2.ParameterName = "@surname"; p2.Value = surname; cmd.Parameters.Add(p2);

            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                list.Add(ReadParticipant(reader));
            }
            return list;
        }

        public List<Participant> FindByEmail(string email)
        {
            var list = new List<Participant>();
            using var connection = _dbUtils.GetConnection();
            connection.Open();

            using var cmd = connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM Participant WHERE email = @email";
            var p = cmd.CreateParameter(); p.ParameterName = "@email"; p.Value = email; cmd.Parameters.Add(p);

            using var reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                list.Add(ReadParticipant(reader));
            }
            return list;
        }

        public Participant? SaveParticipant(Participant participant)
        {
            using var connection = _dbUtils.GetConnection();
            connection.Open();

            using var cmd = connection.CreateCommand();
            cmd.CommandText = @"
                INSERT INTO Participant (id_participant, nume, prenume, varsta, email)
                VALUES (@id, @nume, @prenume, @varsta, @email)";
            var p0 = cmd.CreateParameter(); p0.ParameterName = "@id"; p0.Value = participant.IdParticipant; cmd.Parameters.Add(p0);
            var p1 = cmd.CreateParameter(); p1.ParameterName = "@nume"; p1.Value = participant.Nume; cmd.Parameters.Add(p1);
            var p2 = cmd.CreateParameter(); p2.ParameterName = "@prenume"; p2.Value = participant.Prenume; cmd.Parameters.Add(p2);
            var p3 = cmd.CreateParameter(); p3.ParameterName = "@varsta"; p3.Value = int.Parse(participant.Varsta); cmd.Parameters.Add(p3);
            var p4 = cmd.CreateParameter(); p4.ParameterName = "@email"; p4.Value = participant.Email; cmd.Parameters.Add(p4);

            cmd.ExecuteNonQuery();
            return participant;
        }

        private Participant ReadParticipant(IDataReader reader)
        {
            return new Participant(
                reader.GetString(reader.GetOrdinal("nume")),
                reader.GetString(reader.GetOrdinal("prenume")),
                reader.GetInt32(reader.GetOrdinal("varsta")).ToString(),
                reader.GetString(reader.GetOrdinal("email")))
            {
                IdParticipant = reader.GetString(reader.GetOrdinal("id_participant"))
            };
        }
    }
}
